ngraph/core/reference/include/ngraph/runtime/reference/convolution_backprop_data.hpp

```diff
@@ -0,0 +1,309 @@
+                                         INPUT* delta_in,
+                                         const Shape& out_shape,
+                                         const Shape& filter_shape,
+                                         const Shape& in_shape,
+                                         const Strides& in_dilation,
+                                         const Strides& filter_dilation,
+                                         const CoordinateDiff& forward_in_pad_bellow,
+                                         const CoordinateDiff& forward_in_pad_above,
+                                         const Strides& stride)
+            {
+                // Note that we only reverse the spatial dimensions here (loop
+                // starts at 2)
+                std::vector<INPUT> reversed(shape_size(filter_shape));
+                AxisSet reverse_axes;
+                size_t reverse_axes_start = 2;
+                for (size_t i = reverse_axes_start; i < filter_shape.size(); ++i)
+                {
+                    reverse_axes.insert(i);
+                }
+                reverse(reinterpret_cast<const char*>(filter),
+                        reinterpret_cast<char*>(&reversed[0]),
+                        filter_shape,
+                        filter_shape,
+                        reverse_axes,
+                        sizeof(FILTER));
+                size_t filter_out_channel_axis = 1;
+                size_t filter_in_channel_axis = 0;
+
+                // Compute backward pad out pad bellow
+                size_t spatial_dim_count = in_shape.size() - 2;
+
+                CoordinateDiff backward_delta_out_pad_below;
+                backward_delta_out_pad_below.resize(spatial_dim_count);
+
+                for (size_t i = 0; i < spatial_dim_count; i++)
+                {
+                    backward_delta_out_pad_below[i] =
+                        (static_cast<ptrdiff_t>(filter_shape[i + 2]) - 1) * filter_dilation[i] -
+                        forward_in_pad_bellow[i];
+                }
+                // Compute backward pad out pad above
+                CoordinateDiff backward_delta_out_pad_above;
+                backward_delta_out_pad_above.resize(spatial_dim_count);
+
+                for (size_t i = 0; i < spatial_dim_count; i++)
+                {
+                    backward_delta_out_pad_above[i] =
+                        (static_cast<ptrdiff_t>(filter_shape[i + 2]) - 1) * filter_dilation[i] +
+                        ((forward_in_pad_bellow[i] + ((in_shape[i + 2]) - 1) * in_dilation[i] +
+                          forward_in_pad_above[i] -
+                          (static_cast<ptrdiff_t>(filter_shape[i + 2]) - 1) * filter_dilation[i]) %
+                         stride[i]) -
+                        forward_in_pad_above[i];
+                }
+
+                convolution_backprop_impl<OUTPUT, FILTER, INPUT, ACCUMULATION>(
+                    delta_out,
+                    &reversed[0],
```

Comment:
ConvolutionBackpropData will be addressed in different ticket, so do not review it's code - I didn't change it, only moved it to other file because previously it was together with convolution.